### PR TITLE
Convert to the Endpoint pattern

### LINKED ISSUE #3: `server-relation-broken` handler raise exception
hi @johnsca 
I found an error raised from `uniter hook runner` when I ran `juju remove-relation `.

The root cause is `self.conversation` is not accessible for event `server-relation-broken` because  in `RelationBase.conversation` method, the scope var will be `None` when the relation class scope set to `non GLOBAL` - `JUJU_REMOTE_UNIT=""` in this case. 

the logs as below -> 

```
2018-04-18 05:57:22 DEBUG juju.worker.uniter.operation executor.go:60 running operation run relation-broken (3) hook
2018-04-18 05:57:22 DEBUG juju.worker.uniter uniter.go:641 acquire lock "machine-lock" for uniter hook execution
2018-04-18 05:57:22 DEBUG juju.worker.uniter uniter.go:646 lock "machine-lock" acquired
2018-04-18 05:57:22 DEBUG juju.worker.uniter.operation executor.go:91 preparing operation "run relation-broken (3) hook"
2018-04-18 05:57:22 DEBUG juju.worker.uniter.operation executor.go:91 executing operation "run relation-broken (3) hook"
2018-04-18 05:57:22 DEBUG juju.worker.uniter agent.go:17 [AGENT-STATUS] executing: running server-relation-broken hook
2018-04-18 05:57:23 DEBUG worker.uniter.jujuc server.go:181 running hook tool "config-get"
2018-04-18 05:57:23 DEBUG worker.uniter.jujuc server.go:181 running hook tool "juju-log"
2018-04-18 05:57:23 INFO juju-log server:3: Reactive main running for hook server-relation-broken
2018-04-18 05:57:23 DEBUG worker.uniter.jujuc server.go:181 running hook tool "juju-log"
2018-04-18 05:57:23 INFO juju-log server:3: Invoking reactive handler: ../../application-mysql/charm/hooks/relations/mysql/provides.py:42:departed
2018-04-18 05:57:23 DEBUG worker.uniter.jujuc server.go:181 running hook tool "juju-log"
2018-04-18 05:57:23 ERROR juju-log server:3: Hook error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/charms/reactive/__init__.py", line 72, in main
    bus.dispatch(restricted=restricted_mode)
  File "/usr/local/lib/python3.5/dist-packages/charms/reactive/bus.py", line 367, in dispatch
    _invoke(hook_handlers)
  File "/usr/local/lib/python3.5/dist-packages/charms/reactive/bus.py", line 351, in _invoke
    handler.invoke()
  File "/usr/local/lib/python3.5/dist-packages/charms/reactive/bus.py", line 173, in invoke
    self._action(*args)
  File "/var/lib/juju/agents/application-mysql/charm/hooks/relations/mysql/provides.py", line 44, in departed
    conversation = self.conversation()
  File "/usr/local/lib/python3.5/dist-packages/charms/reactive/relations.py", line 434, in conversation
    raise ValueError('Unable to determine default scope: no current hook or global scope')
ValueError: Unable to determine default scope: no current hook or global scope

2018-04-18 05:57:23 DEBUG server-relation-broken self.scope ---->  service {'JUJU_MODEL_UUID': '9a37894d-61ee-4bca-8596-edf821b14e71', 'JUJU_METER_INFO': 'not set', 'JUJU_SLA': 'unsupported', 'no_proxy': '127.0.0.1,::1,localhost', 'JUJU_REMOTE_UNIT': '', 'JUJU_CHARM_DIR': '/var/lib/juju/agents/unit-mysql-0/charm', 'JUJU_MACHINE_ID': '', 'APT_LISTCHANGES_FRONTEND': 'none', 'NO_PROXY': '127.0.0.1,::1,localhost', 'JUJU_AVAILABILITY_ZONE': '', 'JUJU_CONTEXT_ID': 'mysql/0-server-relation-broken-4129177037886487361', 'JUJU_UNIT_NAME': 'mysql/0', 'JUJU_METER_STATUS': 'AMBER', 'CHARM_DIR': '/var/lib/juju/agents/unit-mysql-0/charm', 'JUJU_RELATION_ID': 'server:3', 'JUJU_PRINCIPAL_UNIT': '', 'JUJU_MODEL_NAME': 'testcaas7', 'JUJU_VERSION': '2.4-beta1', 'PATH': '/var/lib/juju/tools/unit-mysql-0:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin', 'JUJU_AGENT_SOCKET': '@/var/lib/juju/agents/unit-mysql-0/agent.socket', 'JUJU_RELATION': 'server', 'JUJU_API_ADDRESSES': '10.10.145.220:17070', 'DEBIAN_FRONTEND': 'noninteractive', 'JUJU_HOOK_NAME': 'server-relation-broken'}
2018-04-18 05:57:23 DEBUG server-relation-broken self._convers ->  {'units': [''], 'namespace': 'server:3', 'scope': None}
2018-04-18 05:57:23 DEBUG server-relation-broken hookenv ---->   <class 'str'> --->  None
2018-04-18 05:57:23 DEBUG server-relation-broken Traceback (most recent call last):
2018-04-18 05:57:23 DEBUG server-relation-broken   File "/var/lib/juju/agents/unit-mysql-0/charm/hooks/server-relation-broken", line 18, in <module>
2018-04-18 05:57:23 DEBUG server-relation-broken     main()
2018-04-18 05:57:23 DEBUG server-relation-broken   File "/usr/local/lib/python3.5/dist-packages/charms/reactive/__init__.py", line 72, in main
2018-04-18 05:57:23 DEBUG server-relation-broken     bus.dispatch(restricted=restricted_mode)
2018-04-18 05:57:23 DEBUG server-relation-broken   File "/usr/local/lib/python3.5/dist-packages/charms/reactive/bus.py", line 367, in dispatch
2018-04-18 05:57:23 DEBUG server-relation-broken     _invoke(hook_handlers)
2018-04-18 05:57:23 DEBUG server-relation-broken   File "/usr/local/lib/python3.5/dist-packages/charms/reactive/bus.py", line 351, in _invoke
2018-04-18 05:57:23 DEBUG server-relation-broken     handler.invoke()
2018-04-18 05:57:23 DEBUG server-relation-broken   File "/usr/local/lib/python3.5/dist-packages/charms/reactive/bus.py", line 173, in invoke
2018-04-18 05:57:23 DEBUG server-relation-broken     self._action(*args)
2018-04-18 05:57:23 DEBUG server-relation-broken   File "/var/lib/juju/agents/application-mysql/charm/hooks/relations/mysql/provides.py", line 44, in departed
2018-04-18 05:57:23 DEBUG server-relation-broken     conversation = self.conversation()
2018-04-18 05:57:23 DEBUG server-relation-broken   File "/usr/local/lib/python3.5/dist-packages/charms/reactive/relations.py", line 434, in conversation
2018-04-18 05:57:23 DEBUG server-relation-broken     raise ValueError('Unable to determine default scope: no current hook or global scope')
2018-04-18 05:57:23 DEBUG server-relation-broken ValueError: Unable to determine default scope: no current hook or global scope
2018-04-18 05:57:23 ERROR juju.worker.uniter.operation runhook.go:114 hook "server-relation-broken" failed: exit status 1
```

I am thinking there are two ways to solve this:

- to set to the scope to `GLOBAL` in `provides.py` but i am not sure if it will broadcast to all relations that mysql has;

- `def departed` handler only listens `server-relation-departed` event but not `server-relation-broken` (I saw there are some cleanup did in non reactive charm https://api.jujucharms.com/charmstore/v5/mysql-58/archive/hooks/db-relation-broken, wondering how to achieve this in reactive charm here)

would you mind to give some advise?

Thank you very much.
Kelvin

### PR DESCRIPTION
**Pull request by johnsca**
In addition to cleaning up the code, this fixes #3

### CODE REVIEW

**Comment by johnsca** [comment]
Ran in to juju-solutions/charms.reactive#171 with this, but updated to address.  Once that is resolved and released, the handlers can be optimized a bit so that they don't constantly re-run, though it doesn't hurt anything if they do.

File: provides.py [hunk 1]
```diff
@@ -11,76 +11,65 @@
...
+        This returns a mapping of request IDs to application names.
 
-        # if these were requested but not yet fulfilled, cancel the request
-        conversation.remove_state('{relation_name}.database.requested')
+        Example usage::
+
+            @when('clients.database.requested')
+            def create_dbs():
+                mysql = endpoint_from_flag('clients.database.requested')
+                for request, application in mysql.database_requests():
```
Comment (ycliuhw) [inline provides.py]: it would be good to add `mysql.database_requests().items()` here.

Comment (wallyworld) [reply provides.py]: s/good/mandatory since the Python in the example is faulty without the .items() :-)

**Review by wallyworld — APPROVED** [review]
Tested with a gitlab and mysql charm on CAAS and relation now works as expected. Thank you!

**Review by ycliuhw — APPROVED** [review]

**Comment by ycliuhw** [comment]
Thank you very much, @johnsca

**Pull request merged**